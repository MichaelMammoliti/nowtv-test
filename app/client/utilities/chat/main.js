import moment from 'moment';

const normaliseUserCollection = userCollection => {
  return userCollection.reduce((acc, userItem) => {
    acc[userItem.id] = userItem;

    return acc;
  }, {});
};

const sortMessagesByDate = messages => {
  const sortedMessages = messages.sort((a, b) => {
    const currentItemTimestamp = +moment(a.timestamp).format('X');
    const nextItemTimestamp = +moment(b.timestamp).format('X');

    return currentItemTimestamp - nextItemTimestamp;
  });

  return sortedMessages;
};

const extendMessagesWithUserData = (messageCollection, userCollection) => {
  return messageCollection.reduce((acc, messageItem) => {
    const sender = userCollection[messageItem.userId];

    if (sender) {
      const newMessage = {
        ...messageItem,
        sender,
      };

      delete newMessage.userId;

      acc.push(newMessage);
    }

    return acc;
  }, []);
};

export default {
  sortMessagesByDate,
  normaliseUserCollection,
  extendMessagesWithUserData,
};
