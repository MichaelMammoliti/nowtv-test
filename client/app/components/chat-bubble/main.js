import React from 'react';
import moment from 'moment';

import styles from './main.scss';

const ChatBubble = ({ sender, message, timestamp }) => {
  const { avatar, email } = sender;
  const date = (!!timestamp)
    ? moment(timestamp).format('DD MMM YYYY - HH:mm')
    : ''
  ;

  return (
    <div className={styles['chat-bubble']}>

      <div className={styles['chat-bubble__body']}>
        {avatar &&
          <div className={styles['chat-bubble__avatar']}>
            <img src={avatar} alt='profile' />
          </div>
        }

        {message &&
          <div className={styles['chat-bubble__message']}>
            <span>{message}</span>
          </div>
        }
      </div>

      {date &&
        <div className={styles['chat-bubble__date']}>
          <span>{date}</span>
        </div>
      }

      {email &&
        <div className={styles['chat-bubble__email']}>
          <span>{email}</span>
        </div>
      }
    </div>
  );
};

ChatBubble.displayName = 'ChatBubble';

export default ChatBubble;
