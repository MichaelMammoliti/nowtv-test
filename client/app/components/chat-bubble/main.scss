.chat-bubble {
  position: relative;
  padding: 10px;
  background: #fafafa;

  &__body {
    display: flex;
  }

  &__message {
    flex: 1 1 auto;
  }

  &__email {
    position: absolute;
    display: table;
    top: 0;
    left: 50%;
    transform: translate(-50%, -100%);
    background: #000;
    font-size: 14px;
    color: #fff;
    display: none;

    &:before {
      display: block;
      content: '';
      position: absolute;
      display: table;
      bottom: 0;
      left: 50%;
      transform: translate(50%, 50%);
      background: #000;
    }
  }

  &__avatar {
    flex: 0 1 auto;
    padding-right: 20px;

    img {
      border-radius: 50%;
      overflow: hidden;
      height: 40px;
      width: 40px;
    }
  }

  &:hover {
    background-color: #ddd;
  }

  &:hover & {
    &__email {
      display: block;
      padding: 10px;
      border-radius: 10px;
    }
  }
}
