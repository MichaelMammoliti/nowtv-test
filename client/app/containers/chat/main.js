import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import actions from './state/actions';
import styles from './main.scss';

import ChatBubble from '../../components/chat-bubble/main.js';

class Chat extends React.Component {
  componentDidMount() {
    this.props.fetchData();
  }

  render() {
    const { messages, fetchDataRequestStatus } = this.props;

    return (
      <div className={styles['chat']}>
        {fetchDataRequestStatus === 'error' &&
          <div className={styles['chat__error']}>
            <span>cannot render messages.</span>
          </div>
        }

        {fetchDataRequestStatus === 'pending' &&
          <div className={styles['chat__spinner']}>
            <span>fetching data..</span>
          </div>
        }

        {fetchDataRequestStatus === 'success' && messages.length &&
          <div className={styles['chat__body']}>
            {messages.map((messageItem, index) => (
              <div className={styles['chat__body-item']} key={index}>
                <ChatBubble {...messageItem} />
              </div>
            ))}
          </div>
        }
      </div>
    );
  }
}

Chat.displayName = 'Chat';

Chat.defaultProps = {
  messages: [],
};


export {
  Chat,
};

const mapStateToProps = ({ chat }) => ({ ...chat });
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Chat);
