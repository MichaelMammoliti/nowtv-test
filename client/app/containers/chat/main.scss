.chat {
  display: block;
  padding: 40px;

  &__error,
  &__spinner,
  &__body {
    display: block;
  }

  &__body-item {
    margin-bottom: 20px;

    &:last-of-type {
      margin-bottom: 0;
    }
  }
}


