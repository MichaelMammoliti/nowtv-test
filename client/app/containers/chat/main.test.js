import React from 'react';
import { shallow, mount } from 'enzyme';
import { Chat } from './main.js';

const getProps = () => {
  return {
    fetchData: jest.fn(),
  };
};

const getUnit = (props = getProps()) =>
  shallow(<Chat {...props} />)
;

describe('Chat', () => {
  it('should render correctly', () => {
    const unit = getUnit();

    expect(unit).toMatchSnapshot();
  });

  it('should call this.props.fetchData when it mounts', () => {
    const props = getProps();
    const unit = getUnit(props);

    expect(props.fetchData).toHaveBeenCalled();
  });

  it(`should render "ChatBubble" for each item in the "message" prop`, () => {
    const props = getProps();
    props.fetchDataRequestStatus = 'success';
    props.messages = [ {}, {}, {} ];

    const unit = getUnit(props);

    expect(unit.find('ChatBubble').length).toBe(3);
  });

  it(`should render a spinner when the props fetchDataRequestStatus is "pending"`, () => {
    const props = getProps();
    props.fetchDataRequestStatus = 'pending';

    const unit = getUnit(props);

    expect(unit.find('.chat__spinner').exists()).toBe(true);
    expect(unit.find('.chat__error').exists()).toBe(false);
    expect(unit.find('.chat__body').exists()).toBe(false);
  });

  it(`should render "chat-body" when the "fetchDataRequestStatus" prop is "success" and we have messages`, () => {
    const props = getProps();
    props.fetchDataRequestStatus = 'success';
    props.messages = [{}];

    const unit = getUnit(props);

    expect(unit.find('.chat__spinner').exists()).toBe(false);
    expect(unit.find('.chat__error').exists()).toBe(false);
    expect(unit.find('.chat__body').exists()).toBe(true);
  });

  it(`should render an error when the "fetchDataRequestStatus" prop is "fail"`, () => {
    const props = getProps();
    props.fetchDataRequestStatus = 'error';

    const unit = getUnit(props);

    expect(unit.find('.chat__spinner').exists()).toBe(false);
    expect(unit.find('.chat__error').exists()).toBe(true);
    expect(unit.find('.chat__body').exists()).toBe(false);
  });
});