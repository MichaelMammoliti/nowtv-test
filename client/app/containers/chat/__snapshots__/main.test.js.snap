// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Chat
    fetchData={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    messages={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        false,
        false,
      ],
      "className": "chat",
    },
    "ref": null,
    "rendered": Array [
      false,
      false,
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          false,
          false,
        ],
        "className": "chat",
      },
      "ref": null,
      "rendered": Array [
        false,
        false,
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
